### PR TITLE
Better handle thrown exceptions in middlewares

### PR DESCRIPTION
**Pull request by brophdawg11**
Solves some scenarios were exceptions thrown in middlewares prior to the promise chains starting were not caught

### CODE REVIEW

File: src/entry-client.js [hunk 1]
```diff
@@ -86,7 +86,8 @@ export default function initializeClient(createApp, clientOpts) {
                 router,
                 store,
             });
-            return opts.middleware(to, from, store)
+            return Promise.resolve()
+                .then(() => opts.middleware(to, from, store))
```
Comment (brophdawg11) [inline src/entry-client.js]: Using this pattern exceptions throw in middleware will be caught by the promise chain

File: src/entry-server.js [hunk 2]
```diff
@@ -10,8 +10,9 @@ export default function initializeServer(createApp, serverOpts) {
         postMiddleware: () => Promise.resolve(),
     }, serverOpts);
 
-    return context => new Promise((resolve, reject) => {
-        opts.preMiddleware(context).then(() => {
+    return context => new Promise((resolve, reject) => Promise.resolve()
+        .then(() => opts.preMiddleware(context))
```
Comment (brophdawg11) [inline src/entry-server.js]: Same pattern here

File: src/entry-server.js [hunk 3]
```diff
@@ -47,7 +51,8 @@ export default function initializeServer(createApp, serverOpts) {
                 });
 
                 // Execute all provided middleware prior to fetchData
-                return opts.middleware(context, app, router, store)
+                return Promise.resolve()
+                    .then(() => opts.middleware(context, app, router, store))
```
Comment (brophdawg11) [inline src/entry-server.js]: And here

File: src/entry-server.js [hunk 4]
```diff
@@ -69,6 +74,9 @@ export default function initializeServer(createApp, serverOpts) {
                 opts.logger.error('Router rejected onReacy callback');
                 return reject(e);
             });
-        });
-    });
+        })
+        .catch((e) => {
+            opts.logger.error('Error in preMiddleware chain');
```
Comment (urbnjamesmi1) [inline src/entry-server.js]: Should we also log the error here too?

**Pull request merged**